File: lista-encadeada/lista-encadeada.c
#include <stdio.h>
#include <stdlib.h>

// Estrutura do nó na lista
typedef struct No
{
    int valor;
    struct No *proximo;
} No;

// função para criar um novo nó
No *criarNo(int valor)
{
    No *novo = (No *)malloc(sizeof(No));
    novo->valor = valor;
    novo->proximo = NULL;
    return novo;
}

// inserir no inicio
void inserirInicio(No **inicio, int valor)
{
    No *novo = criarNo(valor);
    novo->proximo = *inicio;
    *inicio = novo;
}

// função para inserir no final da lista

void inserirFinal(No **inicio, int valor)
{
    No *novo = criarNo(valor);

    if (*inicio == NULL)
    {
        *inicio = novo;
        return;
    }

    No *temp = *inicio;
    while (temp->proximo != NULL)
    {
        temp = temp->proximo;
    }

    temp->proximo = novo;
}

// função para remover um valor da lista

void remover(No **inicio, int valor)
{
    No *temp = *inicio;
    No *anterior = NULL;

    while (temp != NULL && temp->valor != valor)
    {
        anterior = temp;
        temp = temp->proximo;
    }

    if (temp == NULL)
    {
        printf("valor %d nao encontrado.\n", valor);
        return;
    }

    if (anterior == NULL)
    {
        *inicio = temp->proximo;
    }
    else
    {
        anterior->proximo = temp->proximo;
    }

    free(temp);
}

// Função para exibir os valores da lista

void exibir(No *inicio)
{
    No *temp = inicio;

    if (temp == NULL)
    {
        printf("Lista vazioa.\n");
        return;
    }

    printf("Lista: ");
    while (temp != NULL)
    {
        printf("%d ", temp->valor);
        temp = temp->proximo;
    }

    printf("\n");
}

// limpar lista

void limparLista(No **inicio)
{
    No *temp = *inicio;
    while (temp != NULL)
    {
        No *prox = temp->proximo;
        free(temp);
        temp = prox;
    }

    *inicio = NULL;
    printf("Lista limpa da memoria.\n");
}
// principal

int main()
{
    No *lista = NULL;

    inserirFinal(&lista, 10);
    inserirFinal(&lista, 20);
    inserirFinal(&lista, 30);
    exibir(lista);

    inserirInicio(&lista, 5);
    exibir(lista);

    remover(&lista, 20);
    exibir(lista);

    limparLista(&lista);
    exibir(lista);

    return 0;
}

File: lista-encadeada/lista-encadeada.go
package main

import (
	"fmt"
)

// Estrutura do nó
type No struct {
	valor   int
	proximo *No
}

// Estrutura da lista
type Lista struct {
	inicio *No
}

// Cria um novo nó
func criarNo(valor int) *No {
	return &No{valor: valor, proximo: nil}
}

// Insere no início da lista
func (l *Lista) InserirInicio(valor int) {
	novo := criarNo(valor)
	novo.proximo = l.inicio
	l.inicio = novo
}

// Insere no final da lista
func (l *Lista) InserirFinal(valor int) {
	novo := criarNo(valor)

	if l.inicio == nil {
		l.inicio = novo
		return
	}

	temp := l.inicio
	for temp.proximo != nil {
		temp = temp.proximo
	}
	temp.proximo = novo
}

// Remove um nó com o valor informado
func (l *Lista) Remover(valor int) {
	if l.inicio == nil {
		fmt.Println("Lista vazia.")
		return
	}

	// Caso o valor esteja no primeiro nó
	if l.inicio.valor == valor {
		l.inicio = l.inicio.proximo
		return
	}

	anterior := l.inicio
	temp := l.inicio.proximo

	for temp != nil && temp.valor != valor {
		anterior = temp
		temp = temp.proximo
	}

	if temp == nil {
		fmt.Printf("Valor %d nao encontrado.\n", valor)
		return
	}

	anterior.proximo = temp.proximo
}

// Exibe todos os elementos da lista
func (l *Lista) Exibir() {
	if l.inicio == nil {
		fmt.Println("Lista vazia.")
		return
	}

	temp := l.inicio
	fmt.Print("Lista: ")
	for temp != nil {
		fmt.Printf("%d ", temp.valor)
		temp = temp.proximo
	}
	fmt.Println()
}

// Limpa toda a lista
func (l *Lista) Limpar() {
	l.inicio = nil
	fmt.Println("Lista limpa da memoria.")
}

// Programa principal
func main() {
	lista := Lista{}

	lista.InserirFinal(10)
	lista.InserirFinal(20)
	lista.InserirFinal(30)
	lista.Exibir()

	lista.InserirInicio(5)
	lista.Exibir()

	lista.Remover(20)
	lista.Exibir()

	lista.Limpar()
	lista.Exibir()
}
